CASSGO-22 CASSGO-73 Changes to Query and Batch to make them safely reusable

joao-r-reis opened the pull request:
# API Changes

### `ExecutableQuery`

`ExecutableQuery` is currently an interface that `Query` and `Batch` implements (and is referenced by `HostSelectionPolicy`). However, it is also used in driver internals so the interface contains private methods which makes it impossible for users to "mock" the interface for testing purposes.

In this PR, `ExecutableQuery` is changed so it contains only public methods and is no longer implemented by `Query` and `Batch`. Now, `ExecutableQuery` is used exclusively as a "hook" in `HostSelectionPolicy`. This does mean that users can no longer attempt to cast an `ExecutableQuery` to `Query` or `Batch` but I've added a new method that provides this functionality (`Statement`).

```
type ExecutableQuery interface {
	GetRoutingKey() ([]byte, error)
	Keyspace() string
	Table() string
	IsIdempotent() bool
	Statement() Statement
}
```

### `Statement` 

This is the new interface that represents (and is implemented by) `Query` and `Batch`. In this PR it's only referenced by `ExecutableQuery`.

```
type Statement interface {
	Iter() *Iter
	Exec() error
}
```

### `internalRequest`

New interface that is used by driver internals to decouple the public API of Query/Batch from the internal API.
When creating an internal request object the driver now copies most if not all of the query/batch properties so that users can submit a Query/Batch for execution and then re-use immediately without having to be concerned about the object being modified by the driver execution. It also makes it less error prone because the driver is free to modify these properties (e.g. page state, consistency, query metrics) without causing a change on the objects that the users are using.

```
type internalRequest interface {
	execute(ctx context.Context, conn *Conn) *Iter
	attempt(keyspace string, end, start time.Time, iter *Iter, host *HostInfo)
	retryPolicy() RetryPolicy
	speculativeExecutionPolicy() SpeculativeExecutionPolicy
	getQueryMetrics() *queryMetrics
	RetryableQuery
	ExecutableQuery
}
```

### Query Metrics (i.e. `Query/Batch.Attempts()`, `Query/Batch.Latency()`)

This functionality makes the API a bit awkward because the user submits a query for execution and then inspects the query object to retrieve the side effects of that execution after it's done. It's a better design (imo) to have this kind of data available in the `Iter` object since this is the type that represents the return value of a query/batch.

Due to this, `AddAttempts()`, `Attempts()`, `Latency()` and `AddLatency()` have been removed from `Query` and `Batch`. `Latency()` and `Attempts()` have been added to `Iter` and a new `Batch.Iter()` method has been added so that users can obtain the `Iter` object when executing a batch (even though it doesn't make a lot of sense conceptually due to the name but `Iter` is what the driver uses to return data about a request so it's not just about "iterating").

### ```queryPool```

`Query` objects were allocated from a `sync.Pool` but I don't see the gain of keeping it especially now that `queryMetrics` have been moved to `Iter`. Also users can create these query objects once and store them if they are worried about memory/GC.

### ``` Query/Batch.GetRoutingKey() ```

Having this method as part of the public API of `Query` and `Batch` makes it difficult to manage from the driver maintainer POV, it's supposed to be an internal method that can be called by implementations of `HostSelectionPolicy` but since `ExecutableQuery` was implemented by `Query` and `Batch` this meant that these two types had to have this method on their public API as well.

With this PR, this method is deprecated on `Query` and `Batch` since these two types no longer implement `ExecutableQuery` so we can keep the `Query`/`Batch` API much simpler.

worryg0d commented on query_executor.go:
```diff
@@ -201,16 +210,385 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
+	}
+	return &batchOptions{
+		Type:                  b.Type,
+		Entries:               b.Entries,
```
Here you should use created `newEntries` instead of `b.Entries`

joao-r-reis replied:
Nice catch, done 👍

worryg0d commented on query_executor.go:
```diff
@@ -201,16 +210,385 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
+func (b *internalBatch) releaseAfterExecution() {
+	// empty, because Batch has no equivalent of Query.Release()
+	// that would race with speculative executions.
+}
```
Since both these methods are removed from the interface you can get rid of them here too

joao-r-reis replied:
Nice catch, done

worryg0d commented on conn.go:
```diff
@@ -1523,32 +1534,35 @@ func (c *Conn) UseKeyspace(keyspace string) error {
 
 		if len(entry.Args) > 0 || entry.binding != nil {
-			info, err := c.prepareStatement(batch.Context(), entry.Stmt, batch.trace)
+			info, err := c.prepareStatement(b.batchOpts.context, entry.Stmt, b.batchOpts.trace)
```
If I'm not missing anything here both `ctx` arg of the method and `b.batchOpts.context` should be have the same context value.
The `ctx` arg is being passed by `queryExecutor.executeQuery()` at line 102  by calling `qry.Context()` which should return the underlying `internalBatch.batchOpts.context`.

worryg0d replied:
The only usage of `ctx` is at line 1634 and that is a recursive call to `executeBatch` on an unprepared Cassandra response so changing `b.batchOpts.context` to `ctx` will make the code less confusing and clearer

joao-r-reis replied:
Agree, done 👍

worryg0d reviewed (commented):
Hello @joao-r-reis,

I spent some time reviewing your PR. Great work! It simplifies public query and batch API which is pretty good.

Removing `queryPool` is a good idea. If anyone really needs this, it can be handled on their application side.

I left some minor comments but overall implementation is well.

I am only a bit concerned about the changes this PR provides to the driver internals, especially the conn part which overlaps with #1822.

joao-r-reis commented:
> I am only a bit concerned about the changes this PR provides to the driver internals, especially the conn part which overlaps with https://github.com/apache/cassandra-gocql-driver/pull/1822.

Yeah I'll have to spend some time rebasing this branch and even writing some tests, for now it's good enough if reviewers can provide feedback on the current state of the PR especially the public API changes

lukasz-antoniak commented on query_executor.go:
```diff
@@ -27,32 +27,43 @@ package gocql
 
+// ExecutableQuery is an interface that represents a query or batch statement that
+// exposes the correct functions for the HostSelectionPolicy to operate correctly.
 type ExecutableQuery interface {
```
What do you think about renaming type to `HostPolicyQuery`?

joao-r-reis replied:
I posted a comment about the renaming of this interface below in a response to James

lukasz-antoniak commented on query_executor.go:
```diff
@@ -201,16 +210,375 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
-	qry.releaseAfterExecution()
+}
+
+type queryOptions struct {
```
Shall we group here all relevant options into sub-types, for example:
```
Paging
    pageSize
    initialPageState
    disableAutoPage
Monitoring
    observer
    trace
Parameters
    values
    binding
WriteTimestamp
    defaultTimestamp
    defaultTimestampValue
Consistency
    initialConsistency
    serialConsistency
FrameOptions
    customPayload
    disableSkipMetadata

stmt
prefetch
rt
spec
context
idempotent
keyspace
skipPrepare
routingKey
```
Grouping is not definitive, it is just an idea.

joao-r-reis replied:
yeah I'll do this

joao-r-reis replied:
On a second thought I'm not sure if I like this idea that much, I'd have to create a few new types when maybe I can just group them a bit more by arranging the way they are declared... Wdyt

joao-r-reis replied:
Something like this:

```
type queryOptions struct {
	stmt string

	// Paging
	pageSize         int
	initialPageState []byte
	disableAutoPage  bool

	// Monitoring
	trace    Tracer
	observer QueryObserver

	// Parameters
	values  []interface{}
	binding func(q *QueryInfo) ([]interface{}, error)

	// Timestamp
	defaultTimestamp      bool
	defaultTimestampValue int64

	// Consistency
	initialConsistency Consistency
	serialCons         SerialConsistency

	// Protocol flag
	disableSkipMetadata bool

	customPayload map[string][]byte
	prefetch      float64
	rt            RetryPolicy
	spec          SpeculativeExecutionPolicy
	context       context.Context
	idempotent    bool
	keyspace      string
	skipPrepare   bool
	routingKey    []byte

	// getKeyspace is field so that it can be overriden in tests
	getKeyspace func() string
}
```

lukasz-antoniak commented on query_executor.go:
```diff
@@ -201,16 +210,375 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
+	qryOpts       *queryOptions
+	pageState     []byte
+	metrics       *queryMetrics
+	refCount      uint32
```
Unused?

joao-r-reis replied:
Good catch, leftover from a copy paste

ribaraka replied:
still persist. but changed its position :sweat_smile:

joao-r-reis replied:
oops, maybe I added it back during rebase, fixed

lukasz-antoniak commented on session.go:
```diff
@@ -727,33 +720,36 @@ func (s *Session) routingKeyInfo(ctx context.Context, stmt string) (*routingKeyI
+// Iter executes a batch operation and returns an Iter object
+// that can be used to access properties related to the execution like Iter.Attempts and Iter.Latency
+func (b *Batch) Iter() *Iter {
+	return b.session.executeBatch(b)
```
I feel that people may execute sequence of `Exec()` and `Iter()`. Shall we make an assertion that only one of the is invoked?

jameshartig replied:
Hmm, I see what you mean but I don't think we can lock them into one or another if we want it to be reusable.

joao-r-reis replied:
> I feel that people may execute sequence of `Exec()` and `Iter()`. Shall we make an assertion that only one of the is invoked?

They can't execute sequence (`Exec().Iter()`) unless they do it like:
```
err := b.Exec()
iter := b.Iter()
``` 

Which I believe is already something that can be done with the current API and it will trigger two requests. We can try to improve documentation if this is a concern.

+1 to what James said

jameshartig commented:
> executing a batch (even though it doesn't make a lot of sense conceptually due to the name but Iter is what the driver uses to return data about a request so it's not just about "iterating").

Note that Yugabyte lets you iterate over the [row status](https://docs.yugabyte.com/preview/api/ycql/batch/#row-status) from a batch operation.

jameshartig commented:
> With this PR, this method is deprecated on Query and Batch since these two types no longer implement ExecutableQuery so we can keep the Query/Batch API much simpler.

Why not just remove it? Seems like any `HostSelectionPolicy` implementation would need to make major changes already.

jameshartig commented:
I don't want to bikeshed about names but `ExecutableQuery` is a bit confusing since it contains `Query` and `Batch`. Did that name come from somewhere? Would `ExecutableStatement` make more sense since it has a statement?

jameshartig commented on conn.go:
```diff
@@ -1371,76 +1376,80 @@ func (c *Conn) executeQuery(ctx context.Context, qry *Query) *Iter {
-		qry.routingInfo.keyspace = info.request.keyspace
-		qry.routingInfo.table = info.request.table
-		qry.routingInfo.mu.Unlock()
+		q.routingInfo.mu.Lock()
```
Does this still need to lock? I don't think it does unless we expect `GetRoutingKey` to be called from separate goroutines?

jameshartig replied:
Also can we just change `routingInfo` to not be a struct anymore? Seems like it only was because...
> // routingInfo is a pointer because Query can be copied and copyable struct can't hold a mutex.

Do we still allow copies of `internalQuery`? If so, maybe we should make a Clone method?

joao-r-reis replied:
> Does this still need to lock? I don't think it does unless we expect GetRoutingKey to be called from separate goroutines?

I think this is still the case due to speculative executions (and possibly retries I'm not sure).

> Also can we just change routingInfo to not be a struct anymore? Seems like it only was because...

>>    // routingInfo is a pointer because Query can be copied and copyable struct can't hold a mutex.

> Do we still allow copies of internalQuery? If so, maybe we should make a Clone method?

I don't think it needs to be a struct anymore, I'll change it.

joao-r-reis replied:
So I made some changes to the `routingInfo` part but I didn't remove the struct... I think it still serves a purpose by making it clear that the mutex is used to protect access against the two fields within the struct. Do you still prefer to remove the struct and just have the mutex and two fields in the request type itself?

jameshartig commented on query_executor.go:
```diff
@@ -201,16 +210,375 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
+}
+
+func newQueryOptions(q *Query) *queryOptions {
+	return &queryOptions{
```
Should we be copying `values`, `customPayload`, `initialPageState`, `routingKey`? it might not be obvious that those are not copied and can't be mutated.

joao-r-reis replied:
I think copying `values` and `customPayload` might be a bit too much due to performance reasons. `initialPageState` should probably be copied for safety and since its length is usually pretty small. `routingKey` I'm not so sure, I think it will be `nil` the vast majority of cases anyway but I can make sure it gets copied as well

jameshartig commented on query_executor.go:
```diff
@@ -201,16 +210,375 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
+}
+
+type batchOptions struct {
+	Type                  BatchType
```
Any reason why some of these are public?

joao-r-reis replied:
The type itself is not public and it is never exposed so it wasn't a deliberate choice, just a result of copy paste. I can make them all private for consistency

jameshartig commented:
What about making the methods `Exec(context.Context)` and `Iter(context.Context) error` so there's no need to shallow copy query in `WithContext`? This is a bigger change though that we might not want to make. Taht said, I think Query would be a bit easier to re-use then. I know this differs from `http.Request` but in that case it might be a symptom of not wanting to break the existing API when context's were introduced.

joao-r-reis commented:
> > With this PR, this method is deprecated on Query and Batch since these two types no longer implement ExecutableQuery so we can keep the Query/Batch API much simpler.
> 
> Why not just remove it? Seems like any `HostSelectionPolicy` implementation would need to make major changes already.

Hmm I don't think `HostSelectionPolicy` implementations have to make major changes after this PR unless they rely on casting the `ExecutableQuery` object into `Query` or `Batch` (and even this is a small change, they just need to call .Statement() before doing so). 

> I don't want to bikeshed about names but ExecutableQuery is a bit confusing since it contains Query and Batch. Did that name come from somewhere? Would ExecutableStatement make more sense since it has a statement?

I kept `ExecutableQuery` name unchanged so `HostSelectionPolicy` implementations can pretty much continue to work without many changes (or even any change at all). cc @lukasz-antoniak because you also brought this up in 1 of your comments.
I'm concerned that renaming this interface will increase the chance users will have to change their code when upgrading the driver but I do agree that the current name isn't good.

joao-r-reis commented:
> What about making the methods `Exec(context.Context)` and `Iter(context.Context) error` so there's no need to shallow copy query in `WithContext`? This is a bigger change though that we might not want to make. Taht said, I think Query would be a bit easier to re-use then. I know this differs from `http.Request` but in that case it might be a symptom of not wanting to break the existing API when context's were introduced.

I like the idea but I think it would be too much to ask of users when upgrading since it would affect every single statement on their app... I'd be down to adding an overload and deprecating `WithContext` though... I just don't know what names would fit for the new Iter/Exec overloads...

worryg0d commented:
> I'm concerned that renaming this interface will increase the chance users will have to change their code when upgrading the driver but I do agree that the current name isn't good.

We can add a type alias to leave the `ExecutableQuery` name for the migration period:
```go
// Deprecated: Will be removed in the future major release.
// Please use Statement instead.
type ExecutableQuery = Statement

type Statement interface {
...
}
```

I tested it on go 1.19.13 and it works fine

worryg0d commented:
> I just don't know what names would fit for the new Iter/Exec overloads...

Well, we can use the common **Context** suffix for those overloads, like it is done in sql package from standard lib - https://pkg.go.dev/database/sql#Conn.QueryContext
https://pkg.go.dev/database/sql#DB.ExecContext

joao-r-reis commented:
I've addressed all PR comments, I'll work on rebasing the branch now and then I'll work on the change mentioned [here](https://github.com/apache/cassandra-gocql-driver/pull/1870#issuecomment-2769579576).

joao-r-reis commented:
Rebase is done, it was a bit more complex than I thought it would be so I'd appreciate if you guys could take a look again @jameshartig @lukasz-antoniak @worryg0d

ribaraka commented on query_executor.go:
```diff
@@ -222,16 +238,452 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
+	getKeyspace func() string
+}
+
+func newQueryOptions(q *Query, ctx context.Context) *queryOptions {
+	var newPageState, newRoutingKey []byte
+	if q.initialPageState != nil {
+		pageState := q.initialPageState
+		newPageState = make([]byte, len(pageState))
+		copy(newPageState, pageState)
+	}
```
`newPageState` seems to be redundant here.

joao-r-reis replied:
Well the reason for this is to ensure that a query execution isn't affected if the user tries to modify the page byte slice after query has been submitted for execution

joao-r-reis replied:
I do wonder if we're just adding a performance penalty to safeguard against something that users won't attempt to do 99% of the times

ribaraka replied:
I mean that `newPageState` is copied but never used. And it doesn't seem the `queryOptions` struct has such a field.

joao-r-reis replied:
oh you're right, I removed this field from this type and moved it up to `internalQuery` but I forgot to remove this part, nice catch. Done

ribaraka commented on query_executor.go:
```diff
@@ -222,16 +238,452 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
+		routingInfo:   &queryRoutingInfo{},
+		session:       batch.session,
+	}
+}
```
`consistency` is missing

joao-r-reis replied:
nice catch, done

ribaraka commented on session.go:
```diff
@@ -743,33 +736,47 @@ func (s *Session) routingKeyInfo(ctx context.Context, stmt string, keyspace stri
+// that can be used to access properties related to the execution like Iter.Attempts and Iter.Latency
+func (b *Batch) IterContext(ctx context.Context) *Iter {
+	return b.session.executeBatch(b, ctx)
+}
```
nitpicks: 
1. The `Iter` method should call the `IterContext`
2. The comment for `IterContext` should start with the method name

joao-r-reis replied:
done

ribaraka commented on query_executor.go:
```diff
@@ -27,33 +27,51 @@ package gocql
+}
 
-	withContext(context.Context) ExecutableQuery
+// Statement is an interface that represents a CQL statement that the driver can execute
+// (currently Query and Batch via Session.Query and Session.Batch)
+type Statement interface {
+	Iter() *Iter
+	Exec() error
+}
```
adding here `IterContext` and `ExecContetx`?

joao-r-reis replied:
good suggestion 👍 done

joao-r-reis commented:
@jameshartig @worryg0d do you guys have some free time to take a look at this again during the next few days?

joao-r-reis commented on CHANGELOG.md:
```diff
@@ -54,7 +54,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
-- Make `Session` immutable by removing setters and associated mutex (CASSGO-23)
+- Make Session immutable by removing setters and associated mutex (CASSGO-23)
+
+- Changes to Query and Batch to make them safely reusable (CASSGO-22)
```
reminder to add CASSGO-73 here

joao-r-reis committed:
Changes to Query and Batch API

Before this change queries were mutated while being executed (the query metrics and the consistency for example).
Instead copy query properties to an internal query object and move query metrics to Iter. This allows users
to reuse Query and Batch objects. Query object pooling was also removed.

Some query and batch properties were not accessible via ObservedBatch and ObservedQuery. Added the original Batch and Query
objects to ObservedBatch and ObservedQuery to fix this.

Patch by João Reis; reviewed by James Hartig and Stanislav Bychkov for CASSGO-22 and CASSGO-73
